felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -1314,13 +1363,171 @@ func (m *bpfEndpointManager) setAcceptLocal(iface string, val bool) error {
 
+func (m *bpfEndpointManager) ensureCtlbDevice() {
+	nl, err := netlinkshim.NewRealNetlink()
+	if err != nil {
+		log.Fatal("Failed to create nelink.")
+	}
+
+	_, err = nl.LinkByName("bpfnatin")
+	if err != nil {
+		nat := &netlink.Veth{
+			LinkAttrs: netlink.LinkAttrs{
+				Name: "bpfnatin",
+			},
+			PeerName: "bpfnatout",
+		}
+		if err := nl.LinkAdd(nat); err != nil {
+			log.Fatal("Failed to add bpfnatin.")
+		}
+		link, err := nl.LinkByName("bpfnatin")
+		if err != nil {
+			log.WithError(err).Fatal("Miss bpfnatin after add.")
+		}
+		if err := nl.LinkSetUp(link); err != nil {
+			log.WithError(err).Fatal("Failed to set bpfnatin up.")
+		}
+		link, err = nl.LinkByName("bpfnatout")
+		if err != nil {
+			log.WithError(err).Fatal("Miss bpfnatout after add.")
+		}
+		if err := nl.LinkSetUp(link); err != nil {
+			log.WithError(err).Fatal("Failed to set bpfnatout up.")
+		}
+	}
+
+	if err := configureInterface("bpfnatin", 4, writeProcSys); err != nil {
+		log.WithError(err).Fatal("Failed to configure bpfnatin parameters.")
+	}
+	if err := configureInterface("bpfnatout", 4, writeProcSys); err != nil {
+		log.WithError(err).Fatal("Failed to configure bpfnatout parameters.")
+	}
+
+	err = m.ensureQdisc("bpfnatin")
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to set qdisc on bpfnatin.")
+	}
+
+	if err := m.dp.setRPFilter("bpfnatin", 0); err != nil {
+		log.WithError(err).Fatalf("Failed to disable RPF on bpfnatin.")
+	}
+	if err := m.dp.setRPFilter("bpfnatout", 0); err != nil {
+		log.WithError(err).Fatalf("Failed to disable RPF on bpfnatout.")
+	}
+
+	log.Info("Created bpfnatin pair.")
+
+	m.rulesMakeSpaceForNATIface(nl)
+
+	m.routeTableSNAT = routetable.New(
+		nil,
+		4,
+		false, // vxlan
+		10*time.Second,
+		nil, // deviceRouteSourceAddress
+		1,
+		true, // removeExternalRoutes
+		m.routeTableSNATIdx,
+		m.opReporter,
+	)
+
+	anyV4, _ := ip.CIDRFromString("0.0.0.0/0")
+	m.routeTableSNAT.RouteUpdate("bpfnatout", routetable.Target{
+		Type: routetable.TargetTypeUnicast,
+		CIDR: anyV4,
+	})
+
+	m.routeRuleSNAT, err = routerule.New(
+		4,
+		0, // routing priority
+		set.From(m.routeTableSNATIdx),
+		routerule.RulesMatchSrcFWMarkTable,
+		routerule.RulesMatchSrcFWMarkTable,
+		10*time.Second,
+		func() (routerule.HandleIface, error) {
+			return netlinkshim.NewRealNetlink()
+		},
+		m.opReporter,
+	)
+	if err != nil {
+		log.WithError(err).Panic("Unexpected error creating rule manager")
+	}
+	m.routeRuleSNAT.SetRule(routerule.NewRule(4, 0).
+		GoToTable(m.routeTableSNATIdx).
+		MatchFWMarkWithMask(tcdefs.MarkSeenToNatIfaceOut, tcdefs.MarkSeenToNatIfaceOut),
+	)
+}
+
+func (m *bpfEndpointManager) rulesMakeSpaceForNATIface(nl netlinkshim.Interface) {
+	rules, err := nl.RuleList(unix.AF_INET)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to list routing rules.")
+	}
+	log.WithField("rules", rules).Debugf("Existing routing rules")
+
+	sort.SliceStable(rules, func(i, j int) bool { return rules[i].Priority < rules[j].Priority })
```

Comment:
I'm a bit confused by the handling of negative numbers.  I thought you said that the priority was _really_ a uint32 under the covers.  Should we just cast to uint32 here then the -1 will become 0xffffffff and sort "naturally"

Seems wrong to move 0xffffffff to position 1 below; shouldn't it stay at "the end"?

Reply:
yeah, the netlink lib can return (and take) -1, it pretty much means "not set", the netlink protocol does it by not including the field in the message. Not sure why I get -1 for the default `local` rule at prio 0. Imo kernel sets the prio and should return the prio set, but the default rule might be special.